onyo/commands/get.py

```diff
@@ -0,0 +1,145 @@
+    if not keys:
+        assets = sorted(
+            assets,
+            key=lambda x: [
+                int(s) if s.isdigit() else s.lower()
+                for s in re.split('([0-9]+)', str(x[0]))],
+            reverse=reverse)
+    else:
+        for key in reversed(keys):
+            assets = sorted(
+                assets,
+                key=lambda x: [
+                    int(s) if s.isdigit() else s.lower() for s in
+                    re.split('([0-9]+)', str(x[1][key]))],
+                reverse=reverse)
+
+    return assets
+
+
+def fill_unset(
+        assets: Generator[tuple[Path, dict[str, str]], None, None],
+        keys: list, unset: str = '<unset>') -> Generator:
+    """
+    If a key is not present for an asset, define it as `unset` (default is
+    `'<unset>'`).
+    """
+    unset_keys = {key: unset for key in keys}
+    for asset, data in assets:
+        yield asset, unset_keys | data
+
+
+def eval_depth(depth: Union[str, int, None], default: int) -> int:
+    return int(depth) if depth else default
+
+
+def eval_filters(
+        filters: list[str], repo: Repo, rich: bool = False) -> list[Filter]:
+    """Create filters and check if there are no duplicate filter keys"""
+    init_filters = []
+    try:
+        init_filters = [Filter(f, repo=repo) for f in filters]
+    except OnyoInvalidFilterError as exc:
+        if rich:
+            console = Console(stderr=True)
+            console.print(f'[red]FAILED[/red] {exc}')
+        else:
+            print(exc, file=sys.stderr)
+        exit(1)
+
+    # ensure there are no duplicate filter keys
+    dupe = [
+        x for x, i in Counter([f.key for f in init_filters]).items() if i > 1]
+    if dupe:
+        if rich:
+            console = Console(stderr=True)
+            console.print(f'[red]FAILED[/red] Duplicate filter keys: {dupe}')
+        else:
+            print(f'Duplicate filter keys: {dupe}', file=sys.stderr)
+        exit(1)
+    return init_filters
+
+
+def eval_keys(k: list[str], defaults: list) -> list[str]:
+    """Remove duplicates from k and return default keys if k is empty"""
+    seen = set()
+    k = [x for x in k if not (x in seen or seen.add(x))]
+    return k if k else defaults
+
+
+def eval_path(paths: list[str], defaults: set[str]) -> set[str]:
+    """Return the default path if no paths are given"""
+    return set(paths) if paths else defaults
+
+
+def get(args, opdir):
+    """
+    Display the requested `key`(s) in tabular form for matching assets.
+    """
+    try:
+        repo = Repo(opdir)
+    except OnyoInvalidRepoError:
+        sys.exit(1)
+
+    # Evaluate and format input
+    keys = eval_keys(args.keys, defaults=repo.default_keys)
+    paths = eval_path(args.path, defaults={'.'})
+    depth = eval_depth(args.depth, default=0)
+    filters = eval_filters(
+        args.filter, repo=repo, rich=not args.machine_readable)
+
+    results = repo.get(
+        keys=set(keys), paths=paths, depth=depth, filters=filters)
```

Comment:
When `repo.get()` is used from the API, it should always take in `keys` as a `set`, because there must be no duplicates and the order is not important (sets are unordered). If you check a few lines below the lines you've quoted, you see that `keys` (as a `list`) is used for both sorting and displaying the columns when printing to console. This is intentional, because here the order of the keys is relevant.

Reply:
I wonder: would it be a problem in general to allow duplicates in `repo.get()`?

Because it feels like it would still make more sense to have just one variable `keys`, order/remove duplicates in the command before that `repo.get()` call and then just work with the correct list rest of the command.
And `repo.get()` could then allow sets and lists, and just return the matching assets either way.

Reply:
`Repo.get()` is in the API, so one can assume that it is also allowed to be used outside of `commands.get`.  When it is used elsewhere, the `set` ensures that keys are unique and immediately communicates that to the user by virtue of how sets work.